File: leetcode/865_Smallest_Subtree_with_all_the_Deepest_Nodes/865_Smallest_Subtree_with_all_the_Deepest_Nodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void dfs(TreeNode* root, int& maxDepth, int curDepth) {
        if (root == NULL) return;
        maxDepth = max(maxDepth, curDepth);
        dfs(root->left, maxDepth, curDepth + 1);
        dfs(root->right, maxDepth, curDepth + 1);
    }
    TreeNode* findLca(TreeNode* root, int maxDepth, int curDepth) {
        if (root == NULL) return NULL;
        if (curDepth == maxDepth) return root;
        TreeNode* left = findLca(root->left, maxDepth, curDepth + 1);
        TreeNode* right = findLca(root->right, maxDepth, curDepth + 1);
        if (left != NULL && right != NULL) return root;
        if (left != NULL) return left;
        if (right != NULL) return right;
        return NULL;
    }
    TreeNode* subtreeWithAllDeepest(TreeNode* root) {
        int maxDepth = 0;
        dfs(root, maxDepth, 1);
        return findLca(root, maxDepth, 1);
    }
};

File: leetcode/865_Smallest_Subtree_with_all_the_Deepest_Nodes/865_Smallest_Subtree_with_all_the_Deepest_Nodes.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public void recursive(TreeNode root, int[] maxDepth, int curDepth, List<TreeNode> maxPath, List<TreeNode> path) {
        path.add(root);
        if (root.left == null && root.right == null) {
            if (curDepth > maxDepth[0]) {
                maxPath.clear();
                for (TreeNode i : path) maxPath.add(i);
                maxDepth[0] = curDepth;
            }
            else if (curDepth == maxDepth[0]) {
                int firstIncommon = 0;
                for (; firstIncommon < maxPath.size(); firstIncommon++) {
                    if (maxPath.get(firstIncommon).val != path.get(firstIncommon).val) break;
                }
                maxPath.subList(firstIncommon, maxPath.size()).clear();
            }
        }
        else {
            if (root.left != null) recursive(root.left, maxDepth, curDepth + 1, maxPath, path);
            if (root.right != null) recursive(root.right, maxDepth, curDepth + 1, maxPath, path);
        }
        
        path.remove(path.size() - 1);
    }
    public TreeNode subtreeWithAllDeepest(TreeNode root) {
        List<TreeNode> path = new ArrayList<>();
        List<TreeNode> maxPath = new ArrayList<>();
        int[] maxDepth = new int[1];
        recursive(root, maxDepth, 1, maxPath, path);
        return maxPath.get(maxPath.size() - 1);
    }
}
